File: apps/backend/src/self-observability.ts
import type { SelfObservabilityRuntimeConfig } from "@cloudgrid/runtime";
import type { MessageBridgeMetricRecord, MessageBridgeMetricsRecorder } from "./bridge";
import type { GraphQLMetricRecord, GraphQLMetricsRecorder } from "./graphql-metrics";

type MetricRecord = GraphQLMetricRecord | MessageBridgeMetricRecord;
type ExportFetch = (input: string | URL | Request, init?: RequestInit) => Promise<Response>;
type Signal = "metrics" | "traces" | "logs";

export interface SelfObservabilitySpanRecord {
  name: string;
  traceId?: string;
  spanId?: string;
  parentSpanId?: string;
  traceState?: string;
  attributes?: Record<string, string>;
  result: "success" | "error";
  durationSeconds: number;
}

export interface TraceContext {
  traceId: string;
  spanId: string;
  parentSpanId?: string;
  traceState?: string;
}

export interface SelfObservabilityLogRecord {
  event: string;
  severity: "INFO" | "WARN" | "ERROR";
  attributes?: Record<string, string>;
}

export interface SelfObservabilityTraceRecorder {
  recordSpan(record: SelfObservabilitySpanRecord): void;
}

export interface SelfObservabilityLogRecorder {
  recordLog(record: SelfObservabilityLogRecord): void;
}

interface OTLPSelfObservabilityExporterOptions {
  serviceName: "cloudgrid.bff";
  deploymentMode: string;
  selfObservability: SelfObservabilityRuntimeConfig;
  fetch?: ExportFetch;
  now?: () => Date;
  idGenerator?: () => string;
  maxBuffer?: number;
}

export class OTLPSelfObservabilityExporter
  implements
    GraphQLMetricsRecorder,
    MessageBridgeMetricsRecorder,
    SelfObservabilityTraceRecorder,
    SelfObservabilityLogRecorder
{
  readonly #endpointBase: string;
  readonly #bearerToken: string | undefined;
  readonly #resource: Record<string, string>;
  readonly #fetch: ExportFetch;
  readonly #now: () => Date;
  readonly #idGenerator: () => string;
  readonly #maxBuffer: number;
  readonly #enabled: Record<Signal, boolean>;
  readonly #metricsBuffer: MetricRecord[] = [];
  readonly #spanBuffer: Array<SelfObservabilitySpanRecord & { observedAt: Date }> = [];
  readonly #logBuffer: Array<SelfObservabilityLogRecord & { observedAt: Date }> = [];
  readonly #timer: ReturnType<typeof setInterval>;
  #closed = false;

  constructor(options: OTLPSelfObservabilityExporterOptions) {
    const self = options.selfObservability;
    if (!self.otlpEndpoint) {
      throw new Error(
        "ERR-009 CONFIG_INVALID: CLOUDGRID_SELF_OBSERVABILITY_OTLP_ENDPOINT is required",
      );
    }
    this.#endpointBase = self.otlpEndpoint.replace(/\/$/, "");
    this.#bearerToken = self.otlpBearerToken;
    this.#fetch = options.fetch ?? fetch;
    this.#now = options.now ?? (() => new Date());
    this.#idGenerator = options.idGenerator ?? (() => crypto.randomUUID().replaceAll("-", ""));
    this.#maxBuffer = options.maxBuffer ?? 1024;
    this.#enabled = {
      metrics: self.enabled && self.metricsEnabled,
      traces: self.enabled && self.tracesEnabled,
      logs: self.enabled && self.logsEnabled,
    };
    this.#resource = {
      "service.name": options.serviceName,
      "service.namespace": "cloudgrid",
      "cloudgrid.deployment_mode": options.deploymentMode,
      "cloudgrid.self_observability.project_id": self.projectId,
    };
    if (self.companyId) {
      this.#resource["cloudgrid.self_observability.company_id"] = self.companyId;
    }
    this.#timer = setInterval(() => {
      this.flush().catch(() => {});
    }, self.exportIntervalSeconds * 1000);
    this.#timer.unref?.();
  }

  static fromConfig(
    options: OTLPSelfObservabilityExporterOptions,
  ): OTLPSelfObservabilityExporter | undefined {
    const self = options.selfObservability;
    if (!self.enabled || (!self.metricsEnabled && !self.tracesEnabled && !self.logsEnabled)) {
      return undefined;
    }
    return new OTLPSelfObservabilityExporter(options);
  }

  record(record: MetricRecord): void {
    if (!this.#enabled.metrics) {
      return;
    }
    this.#append(this.#metricsBuffer, record);
  }

  recordSpan(record: SelfObservabilitySpanRecord): void {
    if (!this.#enabled.traces) {
      return;
    }
    this.#append(this.#spanBuffer, { ...record, observedAt: this.#now() });
  }

  recordLog(record: SelfObservabilityLogRecord): void {
    if (!this.#enabled.logs) {
      return;
    }
    this.#append(this.#logBuffer, { ...record, observedAt: this.#now() });
  }

  async flush(): Promise<void> {
    const metrics = this.#metricsBuffer.splice(0);
    const spans = this.#spanBuffer.splice(0);
    const logs = this.#logBuffer.splice(0);

    await Promise.all([
      this.#exportSignal("metrics", metrics, () => this.#metricsPayload(metrics)),
      this.#exportSignal("traces", spans, () => this.#tracesPayload(spans)),
      this.#exportSignal("logs", logs, () => this.#logsPayload(logs)),
    ]);
  }

  async shutdown(): Promise<void> {
    this.#closed = true;
    clearInterval(this.#timer);
    await this.flush();
  }

  #append<T>(buffer: T[], record: T) {
    if (this.#closed || buffer.length >= this.#maxBuffer) {
      return;
    }
    buffer.push(record);
  }

  async #exportSignal(signal: Signal, records: unknown[], payload: () => unknown) {
    if (!this.#enabled[signal] || records.length === 0) {
      return;
    }
    const headers: Record<string, string> = {
      "content-type": "application/json",
    };
    if (this.#bearerToken) {
      headers.authorization = `Bearer ${this.#bearerToken}`;
    }
    try {
      const response = await this.#fetch(`${this.#endpointBase}/v1/${signal}`, {
        method: "POST",
        headers,
        body: JSON.stringify(payload()),
      });
      if (!response.ok) {
        return;
      }
    } catch {
      return;
    }
  }

  #metricsPayload(records: MetricRecord[]) {
    return {
      resourceMetrics: [
        {
          resource: { attributes: otlpAttributes(this.#resource) },
          scopeMetrics: [
            {
              scope: { name: "cloudgrid.self_observability" },
              metrics: records.map((record) => otlpMetric(record, this.#now())),
            },
          ],
        },
      ],
    };
  }

  #tracesPayload(records: Array<SelfObservabilitySpanRecord & { observedAt: Date }>) {
    return {
      resourceSpans: [
        {
          resource: { attributes: otlpAttributes(this.#resource) },
          scopeSpans: [
            {
              scope: { name: "cloudgrid.self_observability" },
              spans: records.map((record) => otlpSpan(record, this.#idGenerator)),
            },
          ],
        },
      ],
    };
  }

  #logsPayload(records: Array<SelfObservabilityLogRecord & { observedAt: Date }>) {
    return {
      resourceLogs: [
        {
          resource: { attributes: otlpAttributes(this.#resource) },
          scopeLogs: [
            {
              scope: { name: "cloudgrid.self_observability" },
              logRecords: records.map((record) => otlpLogRecord(record)),
            },
          ],
        },
      ],
    };
  }
}

export class OTLPMetricsExporter extends OTLPSelfObservabilityExporter {
  static fromConfig(
    options: OTLPSelfObservabilityExporterOptions,
  ): OTLPMetricsExporter | undefined {
    const self = options.selfObservability;
    if (!self.enabled || !self.metricsEnabled) {
      return undefined;
    }
    return new OTLPMetricsExporter({
      ...options,
      selfObservability: { ...self, tracesEnabled: false, logsEnabled: false },
    });
  }
}

function otlpMetric(record: MetricRecord, now: Date) {
  const point: Record<string, unknown> = {
    timeUnixNano: unixNano(now),
    attributes: otlpAttributes(record.attributes as unknown as Record<string, string>),
  };
  if (record.kind === "histogram") {
    return {
      name: record.metric,
      histogram: {
        aggregationTemporality: "AGGREGATION_TEMPORALITY_DELTA",
        dataPoints: [
          { ...point, count: "1", sum: record.value, bucketCounts: ["1"], explicitBounds: [] },
        ],
      },
    };
  }
  return {
    name: record.metric,
    sum: {
      aggregationTemporality: "AGGREGATION_TEMPORALITY_DELTA",
      isMonotonic: true,
      dataPoints: [{ ...point, asDouble: record.value }],
    },
  };
}

function otlpSpan(
  record: SelfObservabilitySpanRecord & { observedAt: Date },
  idGenerator: () => string,
) {
  const traceId = validTraceId(record.traceId) ?? validTraceId(fixedHex(idGenerator(), 32));
  const spanId = validSpanId(record.spanId) ?? validSpanId(fixedHex(idGenerator(), 16));
  const parentSpanId = validSpanId(record.parentSpanId);
  const endUnixNano = BigInt(record.observedAt.getTime()) * 1_000_000n;
  const durationNanos = BigInt(Math.max(0, Math.round(record.durationSeconds * 1_000_000_000)));
  const span: Record<string, unknown> = {
    traceId: hexBytesToBase64(traceId ?? randomHex(32)),
    spanId: hexBytesToBase64(spanId ?? randomHex(16)),
    name: record.name,
    kind: "SPAN_KIND_INTERNAL",
    startTimeUnixNano: `${endUnixNano - durationNanos}`,
    endTimeUnixNano: `${endUnixNano}`,
    attributes: otlpAttributes(record.attributes ?? {}),
    status: { code: record.result === "success" ? "STATUS_CODE_OK" : "STATUS_CODE_ERROR" },
  };
  if (parentSpanId) {
    span.parentSpanId = hexBytesToBase64(parentSpanId);
  }
  if (record.traceState) {
    span.traceState = record.traceState;
  }
  return span;
}

function otlpLogRecord(record: SelfObservabilityLogRecord & { observedAt: Date }) {
  return {
    timeUnixNano: unixNano(record.observedAt),
    severityText: record.severity,
    body: { stringValue: record.event },
    attributes: otlpAttributes(record.attributes ?? {}),
  };
}

function otlpAttributes(attributes: Record<string, string>) {
  return Object.entries(attributes)
    .filter(([key, value]) => key !== "" && value !== "")
    .map(([key, value]) => ({ key, value: { stringValue: value } }));
}

function unixNano(date: Date): string {
  return `${BigInt(date.getTime()) * 1_000_000n}`;
}

function fixedHex(value: string, length: number): string {
  const hex = value
    .replaceAll("-", "")
    .toLowerCase()
    .replace(/[^0-9a-f]/g, "");
  return hex.padEnd(length, "0").slice(0, length);
}

function hexBytesToBase64(value: string): string {
  return Buffer.from(value, "hex").toString("base64");
}

export function createTraceContext(
  options: {
    traceId?: () => string;
    spanId?: () => string;
    parentSpanId?: string;
    traceState?: string;
  } = {},
): TraceContext {
  const parentSpanId = normalizeHex(options.parentSpanId, 16);
  const validParentSpanId = validSpanId(parentSpanId);
  return {
    traceId: validTraceId(fixedHex(options.traceId?.() ?? randomHex(32), 32)) ?? randomHex(32),
    spanId: validSpanId(fixedHex(options.spanId?.() ?? randomHex(16), 16)) ?? randomHex(16),
    ...(validParentSpanId ? { parentSpanId: validParentSpanId } : {}),
    ...(validTraceState(options.traceState) ? { traceState: options.traceState } : {}),
  };
}

export function parseTraceContext(headers: {
  traceparent?: string | null;
  tracestate?: string | null;
}): TraceContext | undefined {
  const traceparent = headers.traceparent?.trim();
  const match = traceparent?.match(/^00-([0-9a-f]{32})-([0-9a-f]{16})-[0-9a-f]{2}$/);
  if (!match) {
    return undefined;
  }
  const traceId = validTraceId(match[1]);
  const parentSpanId = validSpanId(match[2]);
  if (!traceId || !parentSpanId) {
    return undefined;
  }
  return {
    traceId,
    spanId: parentSpanId,
    ...(validTraceState(headers.tracestate) ? { traceState: headers.tracestate } : {}),
  };
}

export function traceContextToTraceParent(context: TraceContext): string {
  return `00-${context.traceId}-${context.spanId}-01`;
}

function randomHex(length: number): string {
  const bytes = new Uint8Array(Math.ceil(length / 2));
  crypto.getRandomValues(bytes);
  const hex = Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0"))
    .join("")
    .slice(0, length);
  if (/^0+$/.test(hex)) {
    return `1${"0".repeat(length - 1)}`;
  }
  return hex;
}

function normalizeHex(value: string | undefined, length: number): string | undefined {
  if (!value) {
    return undefined;
  }
  return fixedHex(value, length);
}

function validTraceId(value: string | undefined): string | undefined {
  return value && /^[0-9a-f]{32}$/.test(value) && !/^0+$/.test(value) ? value : undefined;
}

function validSpanId(value: string | undefined): string | undefined {
  return value && /^[0-9a-f]{16}$/.test(value) && !/^0+$/.test(value) ? value : undefined;
}

function validTraceState(value: string | null | undefined): value is string {
  return typeof value === "string" && value.length <= 512 && /^[\x20-\x7e]*$/.test(value);
}

File: core/control-plane/internal/self_observability.go
package internal

import (
	"bytes"
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"log/slog"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	contracts "github.com/cloudgrid-dev/cloudgrid/core/go-contracts"
	"github.com/cloudgrid-dev/cloudgrid/core/go-runtime/selfobs"
	collectorlogspb "go.opentelemetry.io/proto/otlp/collector/logs/v1"
	collectortracepb "go.opentelemetry.io/proto/otlp/collector/trace/v1"
	commonpb "go.opentelemetry.io/proto/otlp/common/v1"
	logspb "go.opentelemetry.io/proto/otlp/logs/v1"
	resourcepb "go.opentelemetry.io/proto/otlp/resource/v1"
	tracepb "go.opentelemetry.io/proto/otlp/trace/v1"
	"google.golang.org/protobuf/encoding/protojson"
	"google.golang.org/protobuf/proto"
)

type SelfObservabilityRecorder interface {
	RecordSpan(SelfObservabilitySpan)
	RecordLog(SelfObservabilityLog)
	Flush(context.Context) error
	Shutdown(context.Context) error
}

type SelfObservabilitySpan struct {
	Name         string
	TraceID      string
	SpanID       string
	ParentSpanID string
	TraceState   string
	StartTime    time.Time
	EndTime      time.Time
	Attributes   map[string]string
}

type SelfObservabilityLog struct {
	Timestamp    time.Time
	SeverityText string
	Body         string
	Attributes   map[string]string
}

type SelfObservabilitySignalExporterConfig struct {
	Enabled               bool
	Endpoint              string
	BearerToken           string
	ExportIntervalSeconds int
	ServiceName           string
	DeploymentMode        string
	CompanyID             string
	ProjectID             string
	TracesEnabled         bool
	LogsEnabled           bool
	MaxBuffer             int
	Client                *http.Client
	Logger                *slog.Logger
	Now                   func() time.Time
}

type SelfObservabilitySignalExporter struct {
	tracesEndpoint string
	logsEndpoint   string
	bearerToken    string
	client         *http.Client
	logger         *slog.Logger
	now            func() time.Time
	resource       *resourcepb.Resource
	tracesEnabled  bool
	logsEnabled    bool
	maxBuffer      int
	stop           chan struct{}
	stopOnce       sync.Once

	mu     sync.Mutex
	spans  []SelfObservabilitySpan
	logs   []SelfObservabilityLog
	closed bool
}

func NewSelfObservabilitySignalExporter(config SelfObservabilitySignalExporterConfig) (*SelfObservabilitySignalExporter, error) {
	if !config.Enabled {
		return nil, nil
	}
	tracesEndpoint, err := selfObservabilitySignalEndpoint(config.Endpoint, "/v1/traces")
	if err != nil {
		return nil, err
	}
	logsEndpoint, err := selfObservabilitySignalEndpoint(config.Endpoint, "/v1/logs")
	if err != nil {
		return nil, err
	}
	client := config.Client
	if client == nil {
		client = &http.Client{Timeout: 2 * time.Second}
	}
	now := config.Now
	if now == nil {
		now = func() time.Time { return time.Now().UTC() }
	}
	maxBuffer := config.MaxBuffer
	if maxBuffer <= 0 {
		maxBuffer = 1024
	}
	exporter := &SelfObservabilitySignalExporter{
		tracesEndpoint: tracesEndpoint,
		logsEndpoint:   logsEndpoint,
		bearerToken:    strings.TrimSpace(config.BearerToken),
		client:         client,
		logger:         config.Logger,
		now:            now,
		maxBuffer:      maxBuffer,
		tracesEnabled:  config.TracesEnabled,
		logsEnabled:    config.LogsEnabled,
		stop:           make(chan struct{}),
		resource: &resourcepb.Resource{Attributes: selfObservabilityKeyValues(map[string]string{
			"service.name":                            config.ServiceName,
			"service.namespace":                       "cloudgrid",
			"cloudgrid.deployment_mode":               config.DeploymentMode,
			"cloudgrid.self_observability.company_id": config.CompanyID,
			"cloudgrid.self_observability.project_id": config.ProjectID,
		})},
	}
	interval := time.Duration(config.ExportIntervalSeconds) * time.Second
	if interval <= 0 {
		interval = 10 * time.Second
	}
	go exporter.run(interval)
	return exporter, nil
}

func selfObservabilitySignalEndpoint(base string, suffix string) (string, error) {
	base = strings.TrimSpace(base)
	if base == "" {
		return "", errors.New("ERR-009 CONFIG_INVALID: CLOUDGRID_SELF_OBSERVABILITY_OTLP_ENDPOINT is required")
	}
	parsed, err := url.Parse(base)
	if err != nil || parsed.Scheme == "" || parsed.Host == "" {
		return "", errors.New("ERR-009 CONFIG_INVALID: CLOUDGRID_SELF_OBSERVABILITY_OTLP_ENDPOINT must be a valid URL")
	}
	parsed.Path = strings.TrimRight(parsed.Path, "/") + suffix
	parsed.RawQuery = ""
	parsed.Fragment = ""
	return parsed.String(), nil
}

func (exporter *SelfObservabilitySignalExporter) run(interval time.Duration) {
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for {
		select {
		case <-ticker.C:
			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
			_ = exporter.Flush(ctx)
			cancel()
		case <-exporter.stop:
			return
		}
	}
}

func (exporter *SelfObservabilitySignalExporter) RecordSpan(span SelfObservabilitySpan) {
	if exporter == nil || !exporter.tracesEnabled || strings.TrimSpace(span.Name) == "" {
		return
	}
	exporter.mu.Lock()
	defer exporter.mu.Unlock()
	if exporter.closed || len(exporter.spans) >= exporter.maxBuffer {
		return
	}
	span.Attributes = copyStringMap(span.Attributes)
	exporter.spans = append(exporter.spans, span)
}

func (exporter *SelfObservabilitySignalExporter) RecordLog(log SelfObservabilityLog) {
	if exporter == nil || !exporter.logsEnabled || strings.TrimSpace(log.Body) == "" {
		return
	}
	exporter.mu.Lock()
	defer exporter.mu.Unlock()
	if exporter.closed || len(exporter.logs) >= exporter.maxBuffer {
		return
	}
	log.Attributes = copyStringMap(log.Attributes)
	exporter.logs = append(exporter.logs, log)
}

func (exporter *SelfObservabilitySignalExporter) Flush(ctx context.Context) error {
	if exporter == nil {
		return nil
	}
	spans, logs := exporter.drain()
	if len(spans) > 0 {
		_ = exporter.post(ctx, exporter.tracesEndpoint, exporter.tracePayload(spans))
	}
	if len(logs) > 0 {
		_ = exporter.post(ctx, exporter.logsEndpoint, exporter.logPayload(logs))
	}
	return nil
}

func (exporter *SelfObservabilitySignalExporter) Shutdown(ctx context.Context) error {
	if exporter == nil {
		return nil
	}
	exporter.mu.Lock()
	exporter.closed = true
	exporter.mu.Unlock()
	exporter.stopOnce.Do(func() {
		close(exporter.stop)
	})
	return exporter.Flush(ctx)
}

func (exporter *SelfObservabilitySignalExporter) drain() ([]SelfObservabilitySpan, []SelfObservabilityLog) {
	exporter.mu.Lock()
	defer exporter.mu.Unlock()
	spans := append([]SelfObservabilitySpan(nil), exporter.spans...)
	logs := append([]SelfObservabilityLog(nil), exporter.logs...)
	exporter.spans = nil
	exporter.logs = nil
	return spans, logs
}

func (exporter *SelfObservabilitySignalExporter) tracePayload(spans []SelfObservabilitySpan) proto.Message {
	otlpSpans := make([]*tracepb.Span, 0, len(spans))
	for _, span := range spans {
		otlpSpans = append(otlpSpans, selfObservabilityOTLPSpan(span, exporter.now()))
	}
	return &collectortracepb.ExportTraceServiceRequest{ResourceSpans: []*tracepb.ResourceSpans{{
		Resource: exporter.resource,
		ScopeSpans: []*tracepb.ScopeSpans{{
			Scope: &commonpb.InstrumentationScope{Name: "cloudgrid.self_observability"},
			Spans: otlpSpans,
		}},
	}}}
}

func (exporter *SelfObservabilitySignalExporter) logPayload(logs []SelfObservabilityLog) proto.Message {
	records := make([]*logspb.LogRecord, 0, len(logs))
	for _, log := range logs {
		records = append(records, selfObservabilityOTLPLog(log, exporter.now()))
	}
	return &collectorlogspb.ExportLogsServiceRequest{ResourceLogs: []*logspb.ResourceLogs{{
		Resource: exporter.resource,
		ScopeLogs: []*logspb.ScopeLogs{{
			Scope:      &commonpb.InstrumentationScope{Name: "cloudgrid.self_observability"},
			LogRecords: records,
		}},
	}}}
}

func (exporter *SelfObservabilitySignalExporter) post(ctx context.Context, endpoint string, message proto.Message) error {
	payload, err := protojson.Marshal(message)
	if err != nil {
		return err
	}
	request, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(payload))
	if err != nil {
		return err
	}
	request.Header.Set("Content-Type", "application/json")
	if exporter.bearerToken != "" {
		request.Header.Set("Authorization", "Bearer "+exporter.bearerToken)
	}
	response, err := exporter.client.Do(request)
	if err != nil {
		exporter.logFailure()
		return err
	}
	defer response.Body.Close()
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		exporter.logFailure()
		return fmt.Errorf("self-observability export failed with status %d", response.StatusCode)
	}
	return nil
}

func selfObservabilityOTLPSpan(span SelfObservabilitySpan, fallback time.Time) *tracepb.Span {
	start := span.StartTime
	if start.IsZero() {
		start = fallback
	}
	end := span.EndTime
	if end.IsZero() || end.Before(start) {
		end = start
	}
	traceID, spanID := selfObservabilityIDs("span", span.Name, start)
	if isValidTraceID(span.TraceID) {
		traceID, _ = hex.DecodeString(span.TraceID)
	}
	if isValidSpanID(span.SpanID) {
		spanID, _ = hex.DecodeString(span.SpanID)
	}
	otlpSpan := &tracepb.Span{
		TraceId:           traceID,
		SpanId:            spanID,
		Name:              span.Name,
		Kind:              tracepb.Span_SPAN_KIND_INTERNAL,
		StartTimeUnixNano: uint64(start.UnixNano()),
		EndTimeUnixNano:   uint64(end.UnixNano()),
		Attributes:        selfObservabilityKeyValues(span.Attributes),
	}
	if isValidSpanID(span.ParentSpanID) {
		otlpSpan.ParentSpanId, _ = hex.DecodeString(span.ParentSpanID)
	}
	if traceState := strings.TrimSpace(span.TraceState); traceState != "" && len(traceState) <= 512 {
		otlpSpan.TraceState = traceState
	}
	return otlpSpan
}

func selfObservabilityOTLPLog(log SelfObservabilityLog, fallback time.Time) *logspb.LogRecord {
	timestamp := log.Timestamp
	if timestamp.IsZero() {
		timestamp = fallback
	}
	severity := logspb.SeverityNumber_SEVERITY_NUMBER_INFO
	switch strings.ToUpper(strings.TrimSpace(log.SeverityText)) {
	case "WARN", "WARNING":
		severity = logspb.SeverityNumber_SEVERITY_NUMBER_WARN
	case "ERROR":
		severity = logspb.SeverityNumber_SEVERITY_NUMBER_ERROR
	}
	return &logspb.LogRecord{
		TimeUnixNano:   uint64(timestamp.UnixNano()),
		SeverityNumber: severity,
		SeverityText:   strings.ToUpper(strings.TrimSpace(log.SeverityText)),
		Body:           &commonpb.AnyValue{Value: &commonpb.AnyValue_StringValue{StringValue: log.Body}},
		Attributes:     selfObservabilityKeyValues(log.Attributes),
	}
}

func selfObservabilityIDs(parts ...any) ([]byte, []byte) {
	hash := sha256.Sum256([]byte(fmt.Sprint(parts...)))
	return hash[:16], hash[16:24]
}

func isValidTraceID(value string) bool {
	return isValidLowerHex(value, 32) && strings.Trim(value, "0") != ""
}

func isValidSpanID(value string) bool {
	return isValidLowerHex(value, 16) && strings.Trim(value, "0") != ""
}

func isValidLowerHex(value string, length int) bool {
	if len(value) != length {
		return false
	}
	for _, char := range value {
		if !((char >= '0' && char <= '9') || (char >= 'a' && char <= 'f')) {
			return false
		}
	}
	return true
}

func selfObservabilityKeyValues(labels map[string]string) []*commonpb.KeyValue {
	items := make([]*commonpb.KeyValue, 0, len(labels))
	for key, value := range labels {
		key = strings.TrimSpace(key)
		value = strings.TrimSpace(value)
		if key == "" || value == "" {
			continue
		}
		items = append(items, &commonpb.KeyValue{
			Key:   key,
			Value: &commonpb.AnyValue{Value: &commonpb.AnyValue_StringValue{StringValue: value}},
		})
	}
	return items
}

func (exporter *SelfObservabilitySignalExporter) logFailure() {
	if exporter.logger == nil {
		return
	}
	exporter.logger.Warn("self_observability_export_failed",
		"service", controlPlaneService,
		"event", "self_observability_export_failed",
		"request_id", "",
		"error_id", "ERR-013",
		"error_code", "MESSAGE_BRIDGE_UNAVAILABLE",
	)
}

type InMemorySelfObservabilityRecorder struct {
	mu    sync.Mutex
	spans []SelfObservabilitySpan
	logs  []SelfObservabilityLog
}

func NewInMemorySelfObservabilityRecorder() *InMemorySelfObservabilityRecorder {
	return &InMemorySelfObservabilityRecorder{}
}

func (recorder *InMemorySelfObservabilityRecorder) RecordSpan(span SelfObservabilitySpan) {
	recorder.mu.Lock()
	defer recorder.mu.Unlock()
	span.Attributes = copyStringMap(span.Attributes)
	recorder.spans = append(recorder.spans, span)
}

func (recorder *InMemorySelfObservabilityRecorder) RecordLog(log SelfObservabilityLog) {
	recorder.mu.Lock()
	defer recorder.mu.Unlock()
	log.Attributes = copyStringMap(log.Attributes)
	recorder.logs = append(recorder.logs, log)
}

func (recorder *InMemorySelfObservabilityRecorder) Flush(context.Context) error {
	return nil
}

func (recorder *InMemorySelfObservabilityRecorder) Shutdown(context.Context) error {
	return nil
}

func (recorder *InMemorySelfObservabilityRecorder) Spans() []SelfObservabilitySpan {
	recorder.mu.Lock()
	defer recorder.mu.Unlock()
	return append([]SelfObservabilitySpan(nil), recorder.spans...)
}

func (recorder *InMemorySelfObservabilityRecorder) Logs() []SelfObservabilityLog {
	recorder.mu.Lock()
	defer recorder.mu.Unlock()
	return append([]SelfObservabilityLog(nil), recorder.logs...)
}

func (recorder *InMemorySelfObservabilityRecorder) HasSpan(name string) bool {
	for _, span := range recorder.Spans() {
		if span.Name == name {
			return true
		}
	}
	return false
}

func (recorder *InMemorySelfObservabilityRecorder) HasLog(event string) bool {
	for _, log := range recorder.Logs() {
		if log.Attributes["event"] == event {
			return true
		}
	}
	return false
}

func copyStringMap(value map[string]string) map[string]string {
	if len(value) == 0 {
		return nil
	}
	copied := make(map[string]string, len(value))
	for key, item := range value {
		copied[key] = item
	}
	return copied
}

type selfObservabilityBridgeMessage struct {
	BridgeMessage
	response     []byte
	traceContext selfobs.TraceContext
}

func (message *selfObservabilityBridgeMessage) Respond(response []byte) error {
	message.response = append([]byte(nil), response...)
	return message.BridgeMessage.Respond(response)
}

func (message *selfObservabilityBridgeMessage) TraceContext() (selfobs.TraceContext, bool) {
	return message.traceContext, selfobs.FormatTraceParent(message.traceContext) != ""
}

func adaptBridgeHandlerWithSelfObservability(subject string, handler bridgeMessageHandler, recorder SelfObservabilityRecorder) bridgeMessageHandler {
	if recorder == nil {
		return handler
	}
	return func(msg BridgeMessage) {
		start := time.Now().UTC()
		traceContext := selfobs.NewRootTraceContext()
		if headers, ok := msg.(interface{ Header(string) string }); ok {
			if parent, ok := selfobs.TraceContextFromHeaders(headers); ok {
				traceContext = selfobs.NewChildTraceContext(parent)
			}
		}
		capture := &selfObservabilityBridgeMessage{BridgeMessage: msg, traceContext: traceContext}
		handler(capture)
		requestID, ok, bridgeError := responseObservabilityFields(capture.response)
		result := "success"
		if !ok {
			result = "error"
		}
		recorder.RecordSpan(SelfObservabilitySpan{
			Name:         "nats " + subject,
			TraceID:      traceContext.TraceID,
			SpanID:       traceContext.SpanID,
			ParentSpanID: traceContext.ParentSpanID,
			TraceState:   traceContext.TraceState,
			StartTime:    start,
			EndTime:      time.Now().UTC(),
			Attributes: map[string]string{
				"messaging.system":           "nats",
				"messaging.destination.name": boundedControlSubject(subject),
				"cloudgrid.request_id":       requestID,
				"rpc.method":                 boundedControlSubject(subject),
				"result":                     result,
			},
		})
		if bridgeError != nil {
			recorder.RecordLog(SelfObservabilityLog{
				Timestamp:    time.Now().UTC(),
				SeverityText: "WARN",
				Body:         "control plane NATS handler failed",
				Attributes: map[string]string{
					"event":                "nats_handler_failed",
					"cloudgrid.request_id": requestID,
					"error_id":             boundedControlErrorID(bridgeError.ID),
					"error_code":           boundedControlErrorCode(bridgeError.Code),
				},
			})
		}
	}
}

func responseObservabilityFields(payload []byte) (string, bool, *contracts.BridgeError) {
	var response struct {
		RequestID string                 `json:"requestId"`
		OK        bool                   `json:"ok"`
		Error     *contracts.BridgeError `json:"error,omitempty"`
	}
	if len(payload) == 0 || json.Unmarshal(payload, &response) != nil {
		return "", false, nil
	}
	return response.RequestID, response.OK, response.Error
}

func boundedControlSubject(subject string) string {
	if _, ok := ControlSubjects()[subject]; ok {
		return subject
	}
	return "unknown"
}

func boundedControlErrorID(id string) string {
	switch id {
	case "ERR-001", "ERR-003", "ERR-004", "ERR-006", "ERR-013", "ERR-016":
		return id
	default:
		return "ERR-006"
	}
}

func boundedControlErrorCode(code string) string {
	switch code {
	case "VALIDATION_FAILED", "NOT_FOUND", "TRACE_NOT_FOUND", "STORAGE_UNAVAILABLE", "MESSAGE_BRIDGE_UNAVAILABLE", "FORBIDDEN":
		return code
	default:
		return "STORAGE_UNAVAILABLE"
	}
}
